File: akka-samples/akka-docs-java-lambda/src/test/java/docs/stream/cookbook/RecipeParseLines.java
/**
 *  Copyright (C) 2015 Typesafe <http://typesafe.com/>
 */
package docs.stream.cookbook;

import java.util.Arrays;
import java.util.concurrent.TimeUnit;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.io.Framing;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.testkit.JavaTestKit;
import akka.util.ByteString;
import scala.concurrent.Await;
import scala.concurrent.duration.FiniteDuration;
import scala.runtime.BoxedUnit;

public class RecipeParseLines extends RecipeTest {

  static ActorSystem system;

  @BeforeClass
  public static void setup() {
    system = ActorSystem.create("RecipeLoggingElements");
  }

  @AfterClass
  public static void tearDown() {
    JavaTestKit.shutdownActorSystem(system);
    system = null;
  }

  final Materializer mat = ActorMaterializer.create(system);

  @Test
  public void parseLines() throws Exception {
    final Source<ByteString, BoxedUnit> rawData = Source.from(Arrays.asList(
      ByteString.fromString("Hello World"),
      ByteString.fromString("\r"),
      ByteString.fromString("!\r"),
      ByteString.fromString("\nHello Akka!\r\nHello Streams!"),
      ByteString.fromString("\r\n\r\n")));

    //#parse-lines
    final Source<String, BoxedUnit> lines = rawData
      .via(Framing.delimiter(ByteString.fromString("\r\n"), 100, true))
      .map(b -> b.utf8String());
    //#parse-lines

    Await.result(lines.grouped(10).runWith(Sink.head(), mat), new FiniteDuration(1, TimeUnit.SECONDS));
  }

}

File: akka-docs-dev/rst/scala/code/docs/stream/cookbook/RecipeParseLines.scala
package docs.stream.cookbook

import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source
import akka.util.ByteString

import scala.annotation.tailrec
import scala.concurrent.Await
import scala.concurrent.duration._

class RecipeParseLines extends RecipeSpec {

  "Recipe for parsing line from bytes" must {

    "work" in {
      val rawData = Source(List(
        ByteString("Hello World"),
        ByteString("\r"),
        ByteString("!\r"),
        ByteString("\nHello Akka!\r\nHello Streams!"),
        ByteString("\r\n\r\n")))

      //#parse-lines
      import akka.stream.io.Framing
      val linesStream = rawData.via(Framing.lines("\r\n", maximumLineBytes = 100))
      //#parse-lines

      Await.result(linesStream.grouped(10).runWith(Sink.head), 3.seconds) should be(List(
        "Hello World\r!",
        "Hello Akka!",
        "Hello Streams!",
        ""))
    }

  }

}
